keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,1190 @@
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+The ability to shift left validation logic from controllers to the front-end
+is a long-term goal for improving the useability of the Kubernetes project. As
+it stands today, our treatment of validation for unchanged fields stands as a barrier to both CRD authors and Kubernetes developers to adopting the available validation features.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+CRD authors today are thrown into the deep end when it comes to validation of
+the values in their schemas. Generally authors understand to increment version
+when adjusting the field layout of a CRD, but to do so when only modifying value validations is cumbersome, and more trouble than it is worth.
+
+Modifying a value validation on a CRD today means that you risk breaking the
+workflow of all your users, this high price to pay limits adoption, and degrades the Kubernetes user experience: we are prevented from shifting validation logic left.
+
+Additionally, this sad state of affairs stunts the advancement of Kubernetes itself. [KEP-3937](https://github.com/kubernetes/enhancements/pull/3938) proposes to add declarative validation to Kubernetes, but to
+do that would require adding new `format` types. This would also break Kubernetes user workflows. 
+
+See User Stories for more detailed description of
+the motivation behind this KEP.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+1. Remove barriers blocking CRD authors from widening value validations
+2. Remove barriers blocking CRD authors from tightening value validations
+3. Remove barriers blocking Kubernetes from widening validations
+4. Remove barriers blocking Kubernetes from tightening validations
```

Comment:
Is doing that automatically for all CRDs a goal or an anti-goal?

Reply:
Note: this is technically a breaking change of behaviour.

Reply:
This does change the current behavior which would block an update containing an unchanged field that now fails validation.

I think it is worth the tradeoff. So I would like for it to apply to all CRDs, but open to making it opt-in via extension or at the cluster level feature gate.

Ideas?

/cc @jpbetz

Reply:
May add a line item somewhere to discuss the breaking behavior more before promoting to beta? I don't think we should worry about it much for alpha.

My intuition is that the biggest disadvantage would be to CRD authors that fail to notice they made a breaking change because they were using UPDATE to test if the change was safe and will need to test with CREATE once ratcheting is enabled.